### PR TITLE
PE/COFF Image Validation Tool

### PR DESCRIPTION
**Pull request by Javagedes**
Add a PE/COFF image validation tool for command line that can be used to
verify that memory protection requirements such as section alignment and
write / execute exclusion settings have been applied correctly to
PE/COFF files.

### CODE REVIEW

File: docs/usability/using_image_validation_tool.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    [Optional] Clears the NX_COMPAT flag. returns <file>_nx_clear.<filetype>
+
+### --get-nx-compat
+
+    [Optional] Returns the value of the NX_COMPAT flag
+
+### Status codes returned
+
+    0 - All tests returned pass or warn
+    1 - One ore more tests returned fail, or an error has occurred
```
Comment (makubacki) [inline docs/usability/using_image_validation_tool.md:41]:
```suggestion
    1 - One or more tests returned fail, or an error has occurred
```

File: edk2toolext/image_validation.py [hunk 2]
```diff
@@ -0,0 +1,612 @@
...
+                        ]
+                    },
+                    "DRIVER": {
+                        "ALLOWED_SUBSYSTEMS": [
+                            "IMAGE_SUBSYSTEM_EFI_BOOT_SERVICE_DRIVER",
+                            "IMAGE_SUBSYSTEM_EFI_RUNTIME_DRIVER"
+                            "IMAGE_SUBSYSTEM_EFI_ROM"
+                        ]
+                    },
+                    "PEI": {}
```
Comment (makubacki) [inline edk2toolext/image_validation.py:131]: Is "PEI" required if it is empty?

Comment (Javagedes) [reply edk2toolext/image_validation.py:131]: Technically, no. However, I wanted to leave it in as the other machine types did have specific requirements for PEI. I'm fine with removing it if you think it is best.

File: docs/usability/using_image_validation_tool.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
+# PE/COFF Image Validation
+
+The PE/COFF image validation tool is a command line tool used to verify tha
+memory protection requirements such as section alignment and write / execute
+settings are applied correctly. This tool also provides the ability to check,
+set, and clear the NX_COMPAT flag found in OPTIONAL_HEADER.DllCharacteristics.
+
+## Synopsis
+
+image_validation.py [-h] -i FILE [-d] [-s MODULE] [--set-nx-compat SET_NX_COMPAT] [--clear-nx-compat CLEAR_NX_COMPAT] [--get-nx-compat]
```
Comment (Javagedes) [inline docs/usability/using_image_validation_tool.md:10]:
```suggestion
image_validation.py [-h] -i FILE [-d] [-p PROFILE] [--set-nx-compat] [--clear-nx-compat] [--get-nx-compat]
```

File: docs/usability/using_image_validation_tool.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    [Optional] Provides information on flags
+
+### -i, --file
+
+    [Required] The input PE/COFF file to be verified
+
+### -p, --profile
+
+    [Optional] The profile config to be verified against. Will use the default, if not provided
+
```
Comment (Javagedes) [inline docs/usability/using_image_validation_tool.md:25]:
```suggestion
### -d, --debug

    [Optional] Sets the logging mode to debug
    
```

File: docs/usability/using_image_validation_tool.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
+# PE/COFF Image Validation
+
+The PE/COFF image validation tool is a command line tool used to verify tha
```
Comment (Javagedes) [inline docs/usability/using_image_validation_tool.md:3]:
```suggestion
The PE/COFF image validation tool is a command line tool used to verify that
```

**Review by Javagedes — COMMENTED** [review]
Minor updates to the readme.

File: edk2toolext/image_validation.py (hunk 2 above)
Comment (makubacki) [reply edk2toolext/image_validation.py:131]: Some other sections show its usage and I think it's generally understood that it is not required if it is not used so I'd recommend removing it to reinforce that expectation.

File: docs/usability/using_image_validation_tool.md (hunk 3 above)
Comment (makubacki) [reply docs/usability/using_image_validation_tool.md:10]: Thanks, I was a little confused by the previous parameters.

**Review by makubacki — APPROVED** [review]

File: edk2toolext/image_validation.py [hunk 6]
```diff
@@ -0,0 +1,612 @@
...
+        Write-able or Read-able, but not both.
+    """
+
+    def name(self):
+        return 'Section data / code separation verification'
+
+    def execute(self, pe, config_data):
+        target_requirements = config_data["TARGET_REQUIREMENTS"]
+
+        if (not target_requirements.get("DATA_CODE_SEPARATION")
```
Comment (corthon) [inline edk2toolext/image_validation.py:371]:
```suggestion
        if (not target_requirements.get("DATA_CODE_SEPARATION")
        if target_requirements.get("DATA_CODE_SEPARATION", False) is False:
```

Comment (corthon) [reply edk2toolext/image_validation.py:371]: This didn't format right, but you get the idea.

**Review by corthon — CHANGES_REQUESTED** [review]
On the whole I find this to be very well-written and readable. If we can get a couple things answered we should be able to get this in.

File: edk2toolext/image_validation.py [hunk 7]
```diff
@@ -0,0 +1,612 @@
...
+            Type
+
+        @return Result.PASS : All tests passed successfully (including warnings)
+        @return Result.SKIP : There is no information in the config file for the target and fv file type
+        @return Result.ERROR: At least one test failed. Error messages can be found in the log
+        """
+
+        # Catch any invalid profiles
+        machine_type = MACHINE_TYPE[pe.FILE_HEADER.Machine]
+        if not self.config_data[machine_type].get(profile):
```
[the remark this answers is not in the document]
Comment (Javagedes) [reply edk2toolext/image_validation.py:302]: 1. I had some internal debate against whether I should error proof this section or exit after a bad profile. I went with error proofing. However, I think you are right. Just raising an exception and exiting is probably better. Setting the profile to default could, in the future, lead to a PE/COFF file passing validation when it shouldn't.
2. If I was to stick with the current implementation, which I'm not, then yes, using the `default` parameter for `get()` would be cleaner. Honestly, I didn't know that existed for `get()`!

File: edk2toolext/image_validation.py [hunk 8]
```diff
@@ -0,0 +1,612 @@
...
+                        action='store_true',
+                        default=False)
+
+    parser.add_argument('-p', '--profile',
+                        type=str,
+                        default=None,
+                        help='[Optional] The profile config to be verified againt. \
+                            Will use the default, if not provided')
+
+    group = parser.add_argument_group('NX compatability flag control')
```
[the remark this answers is not in the document]
Comment (Javagedes) [reply edk2toolext/image_validation.py:534]: Originally just a group and if multiple were set, I would do the first and exit. Using mutual exclusion is better here. Will Update.

File: docs/usability/using_image_validation_tool.md (hunk 3 above)
Comment (Javagedes) [inline docs/usability/using_image_validation_tool.md]:
```suggestion
image_validation.py [-h] -i FILE [-d] [-p PROFILE] [--set-nx-compat | --clear-nx-compat | --get-nx-compat]
```

File: docs/usability/using_image_validation_tool.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
+# PE/COFF Image Validation
+
+The PE/COFF image validation tool is a command line tool used to verify that
+memory protection requirements such as section alignment and write / execute
+settings are applied correctly. This tool also provides the ability to check,
+set, and clear the NX_COMPAT flag found in OPTIONAL_HEADER.DllCharacteristics.
+
+## Synopsis
+
+image_validation.py [-h] -i FILE [-d] [-p PROFILE] [--set-nx-compat] [--clear-nx-compat] [--get-nx-compat]
```
Comment (Javagedes) [inline docs/usability/using_image_validation_tool.md:10]:
```suggestion
image_validation.py [-h] -i FILE [-d] [-p PROFILE] [--set-nx-compat | --clear-nx-compat | --get-nx-compat]
```

**Review by corthon — APPROVED** [review]

**Pull request merged**